### PR TITLE
Generate gamestate overlay table with an `include/tables/` header

### PR DESCRIPTION
**Pull request by Thar0**
Also generates the conditions in `Graph_GetNextGameState` for table lookup with the unrelocated address of the init function

### CODE REVIEW

File: src/code/graph.c [hunk 1]
```diff
@@ -113,27 +113,28 @@ void Graph_InitTHGA(GraphicsContext* gfxCtx) {
...
 
+// Used below in `Graph_GetNextGameState`
+#define DEFINE_GAMESTATE_INTERNAL(typeName) GAMESTATE_INDEX_##typeName,
+#define DEFINE_GAMESTATE(typeName, name) DEFINE_GAMESTATE_INTERNAL(typeName)
+typedef enum {
+#include "tables/gamestate_table.h"
+    GAMESTATE_INDEX_MAX
+} GameStateTableIndex;
+#undef DEFINE_GAMESTATE
+#undef DEFINE_GAMESTATE_INTERNAL
```
Comment (AngheloAlf) [inline src/code/graph.c:124]: I would prefer this to be in a header instead

Comment (Thar0) [reply src/code/graph.c:124]: I'm not sure there's really any merit to being in a header besides avoiding potential type clashes. The enum values are not "nice", e.g. the PlayState enum value is `GAMESTATE_ENUM_Play`, it serves no purpose other than to be used in `Graph_GetNextGameState`

**Review by AngheloAlf — APPROVED** [review]

File: src/code/graph.c (hunk 1 above)
Comment (hensldm) [reply src/code/graph.c:124]: While I know it would only get this one use, it still seems nice to have a global enum (pretty sure this would be nice if you wanted to add an `Id` field to the `GameState`). Personally I think it would be nice to add a second/third argument for an enum value that is also defined in the table (like actors), and use that instead of the concatenation. Could also add one for the init function name, but that is even more kind of whatever.

File: include/tables/gamestate_table.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ *    - Argument 3: Gamestate overlay spec segment name
+ *
+ * DEFINE_GAMESTATE_INTERNAL should be used for gamestates that aren't an overlay, the first two arguments are the same as for DEFINE_GAMESTATE
+ */
+/* 0x0000 */ DEFINE_GAMESTATE_INTERNAL(Setup, GAMESTATE_ID_SETUP)
+/* 0x0001 */ DEFINE_GAMESTATE(MapSelect, GAMESTATE_ID_MAP_SELECT, select)
+/* 0x0002 */ DEFINE_GAMESTATE(ConsoleLogo, GAMESTATE_ID_CONSOLE_LOGO, title)
+/* 0x0003 */ DEFINE_GAMESTATE_INTERNAL(Play, GAMESTATE_ID_PLAY)
+/* 0x0004 */ DEFINE_GAMESTATE(TitleSetup, GAMESTATE_ID_TITLE_SETUP, opening)
+/* 0x0005 */ DEFINE_GAMESTATE(FileSelect, GAMESTATE_ID_FILE_SELECT, file_choose)
```
Comment (AngheloAlf) [inline include/tables/gamestate_table.h:16]: I think we could probably drop a few zeroes here

**Review by hensldm — APPROVED** [review]

File: src/code/z_game_dlftbls.c [hunk 3]
```diff
@@ -1,18 +1,25 @@
...
-    GAMESTATE_OVERLAY(select, MapSelect_Init, MapSelect_Destroy, sizeof(MapSelectState)),
-    GAMESTATE_OVERLAY(title, ConsoleLogo_Init, ConsoleLogo_Destroy, sizeof(ConsoleLogoState)),
-    GAMESTATE_OVERLAY_INTERNAL(Play_Init, Play_Destroy, sizeof(PlayState)),
-    GAMESTATE_OVERLAY(opening, TitleSetup_Init, TitleSetup_Destroy, sizeof(TitleSetupState)),
-    GAMESTATE_OVERLAY(file_choose, FileSelect_Init, FileSelect_Destroy, sizeof(FileSelectState)),
+#include "tables/gamestate_table.h"
 };
+
+#undef DEFINE_GAMESTATE
+#undef DEFINE_GAMESTATE_INTERNAL
```
Comment (Roman971) [inline src/code/z_game_dlftbls.c:25]: You may have left it out on purpose, but I believe we should be able to add another set of defines/include at the top of this file to declare the segment symbols required by the table? (similarly to how we have it for `z_actor_dlftbls.c` and such)
This would further automate the segment declaration part, and assuming the symbols are only used in this table we could remove the DECLARE_OVERLAY_SEGMENT lines currently in `segment_symbols.h`

File: include/tables/gamestate_table.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ *    - Argument 3: Gamestate overlay spec segment name
+ *
+ * DEFINE_GAMESTATE_INTERNAL should be used for gamestates that aren't an overlay, the first two arguments are the same as for DEFINE_GAMESTATE
+ */
+/* 0x00 */ DEFINE_GAMESTATE_INTERNAL(Setup, GAMESTATE_ID_SETUP)
+/* 0x01 */ DEFINE_GAMESTATE(MapSelect, GAMESTATE_ID_MAP_SELECT, select)
+/* 0x02 */ DEFINE_GAMESTATE(ConsoleLogo, GAMESTATE_ID_CONSOLE_LOGO, title)
+/* 0x03 */ DEFINE_GAMESTATE_INTERNAL(Play, GAMESTATE_ID_PLAY)
+/* 0x04 */ DEFINE_GAMESTATE(TitleSetup, GAMESTATE_ID_TITLE_SETUP, opening)
+/* 0x05 */ DEFINE_GAMESTATE(FileSelect, GAMESTATE_ID_FILE_SELECT, file_choose)
```
Comment (Roman971) [inline include/tables/gamestate_table.h:16]: Also pretty minor but having _ID in the enum name feels a bit off since we haven't been including it for most other "id" enums (eg. actors, objects, effects)

**Review by Roman971 — COMMENTED** [review]
Looks mostly good to me, just a few comments/notes

File: src/code/graph.c [hunk 5]
```diff
@@ -425,12 +416,12 @@ void Graph_ThreadEntry(void* arg0) {
     GameStateOverlay* ovl;
     char faultMsg[0x50];
 
-    nextOvl = &gGameStateOverlayTable[0];
+    nextOvl = gGameStateOverlayTable;
```
[the remark this answers is not in the document]
Comment (fig02) [reply src/code/graph.c:419]: agree

File: src/code/z_game_dlftbls.c [hunk 6]
```diff
@@ -1,18 +1,35 @@
...
+    { NULL, 0, 0, NULL, NULL, NULL, typeName##_Init, typeName##_Destroy, NULL, NULL, 0, sizeof(typeName##State) },
+
+#define DEFINE_GAMESTATE(typeName, enumName, name) \
+    { NULL,                                        \
+      (uintptr_t)_ovl_##name##SegmentRomStart,     \
+      (uintptr_t)_ovl_##name##SegmentRomEnd,       \
+      _ovl_##name##SegmentStart,                   \
+      _ovl_##name##SegmentEnd,                     \
+      NULL,                                        \
+      typeName##_Init,                             \
```
Comment (Roman971) [inline src/code/z_game_dlftbls.c:23]: It's probably better to leave it for a different PR as it might need some discussion, but I just realized if we're doing this we could also change the `SET_NEXT_GAMESTATE` macro to only take the type name and build the _Init and State names on its own

**Review by Roman971 — APPROVED** [review]
